File: src/year2024/day5/day5.rs
//! [Day 5: Print Queue](https://adventofcode.com/2024/day/5)

use rustc_hash::{FxHashMap, FxHashSet};

struct Puzzle {
    ordering_rules: FxHashMap<i32, FxHashSet<i32>>,
    page_updates: Vec<Vec<i32>>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        let mut ordering_rules: FxHashMap<i32, FxHashSet<i32>> = FxHashMap::default();
        let mut page_updates = Vec::new();

        let data = data.split_once("\n\n").unwrap();

        //
        for line in data.0.lines() {
            let (p1, p2) = line.split_once('|').unwrap();

            let p1 = p1.parse().unwrap();
            let p2 = p2.parse().unwrap();

            ordering_rules.entry(p1).or_default().insert(p2);
        }

        //
        for line in data.1.lines() {
            let v: Vec<_> = line.split(',').map(|x| x.parse::<i32>().unwrap()).collect();
            page_updates.push(v);
        }

        Self {
            ordering_rules,
            page_updates,
        }
    }

    /// Solve part one.
    fn part1(&self) -> i32 {
        let mut result = 0;

        for pu in &self.page_updates {
            let mut correctly_ordered = true;

            for (i, page) in pu.iter().enumerate() {
                if let Some(or) = self.ordering_rules.get(page) {
                    let hs: FxHashSet<i32> = pu.iter().copied().take(i).collect();

                    if hs.intersection(or).count() != 0 {
                        correctly_ordered = false;
                        break;
                    }
                }
            }

            if correctly_ordered {
                result += pu[pu.len() / 2];
            }
        }

        result
    }

    /// Solve part two.
    fn part2(&self) -> i32 {
        self.page_updates
            .iter()
            .map(|x| self.bubble_sort_updates(x))
            .sum()
    }

    /// Sort the page updates.
    /// If a fix has been made, return the middle page number.
    /// Return 0 otherwise.
    fn bubble_sort_updates(&self, updates: &[i32]) -> i32 {
        let mut pu = updates.to_vec();
        let mut reorder = false;
        let mut i = 0;

        while i < pu.len() {
            let page = pu[i];

            if let Some(or) = self.ordering_rules.get(&page) {
                let hs: FxHashSet<i32> = pu.iter().copied().take(i).collect();

                if hs.intersection(or).count() != 0 {
                    pu.swap(i - 1, i);

                    i = 0;
                    reorder = true;
                    continue;
                }
            }

            i += 1;
        }

        if reorder {
            pu[pu.len() / 2]
        } else {
            0
        }
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (i32, i32) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 143);
    }

    #[test]
    fn test02() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 123);
    }
}

File: src/year2024/day5/day5.go
// [Day 5: Print Queue](https://adventofcode.com/2024/day/5)

package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

type Puzzle struct {
	orderingRules map[int]map[int]bool
	pageUpdates   [][]int
}

func newPuzzle(data string) *Puzzle {
	p := &Puzzle{
		orderingRules: make(map[int]map[int]bool),
		pageUpdates:   make([][]int, 0),
	}

	parts := strings.SplitN(data, "\n\n", 2)
	if len(parts) != 2 {
		log.Fatal("Invalid input format")
	}

	// Parse ordering rules
	for _, line := range strings.Split(parts[0], "\n") {
		if line == "" {
			continue
		}
		rule := strings.Split(line, "|")
		if len(rule) != 2 {
			continue
		}
		p1, err1 := strconv.Atoi(strings.TrimSpace(rule[0]))
		p2, err2 := strconv.Atoi(strings.TrimSpace(rule[1]))
		if err1 != nil || err2 != nil {
			continue
		}

		if p.orderingRules[p1] == nil {
			p.orderingRules[p1] = make(map[int]bool)
		}
		p.orderingRules[p1][p2] = true
	}

	// Parse page updates
	for _, line := range strings.Split(parts[1], "\n") {
		if line == "" {
			continue
		}
		parts := strings.Split(line, ",")
		update := make([]int, 0, len(parts))
		for _, part := range parts {
			val, err := strconv.Atoi(strings.TrimSpace(part))
			if err != nil {
				continue
			}
			update = append(update, val)
		}
		if len(update) > 0 {
			p.pageUpdates = append(p.pageUpdates, update)
		}
	}

	return p
}

func (p *Puzzle) part1() int {
	result := 0

	for _, pu := range p.pageUpdates {
		correctlyOrdered := true

		for i, page := range pu {
			if or, exists := p.orderingRules[page]; exists {
				hs := make(map[int]bool)
				for j := 0; j < i; j++ {
					hs[pu[j]] = true
				}

				hasIntersection := false
				for k := range or {
					if hs[k] {
						hasIntersection = true
						break
					}
				}

				if hasIntersection {
					correctlyOrdered = false
					break
				}
			}
		}

		if correctlyOrdered {
			result += pu[len(pu)/2]
		}
	}

	return result
}

func (p *Puzzle) part2() int {
	result := 0
	for _, update := range p.pageUpdates {
		result += p.bubbleSortUpdates(update)
	}
	return result
}

// Sort the page updates.
// If a fix has been made, return the middle page number.
// Return 0 otherwise.
func (p *Puzzle) bubbleSortUpdates(updates []int) int {
	pu := make([]int, len(updates))
	copy(pu, updates)
	reorder := false
	i := 0

	for i < len(pu) {
		page := pu[i]

		if or, exists := p.orderingRules[page]; exists {
			hs := make(map[int]bool)
			for j := 0; j < i; j++ {
				hs[pu[j]] = true
			}

			hasIntersection := false
			for k := range or {
				if hs[k] {
					hasIntersection = true
					break
				}
			}

			if hasIntersection {
				pu[i-1], pu[i] = pu[i], pu[i-1]
				i = 0
				reorder = true
				continue
			}
		}

		i++
	}

	if reorder {
		return pu[len(pu)/2]
	}
	return 0
}

func solve(data []byte) (int, int) {
	input := string(data)
	puzzle := newPuzzle(input)
	return puzzle.part1(), puzzle.part2()
}

func main() {
	filename := "input.txt"
	elapsed := false

	for i := 1; i < len(os.Args); i++ {
		arg := os.Args[i]
		if arg == "--elapsed" {
			elapsed = true
		} else if !strings.HasPrefix(arg, "-") {
			filename = arg
		}
	}

	data, err := os.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}

	if elapsed {
		start := time.Now()
		result1, result2 := solve(data)
		duration := time.Since(start)

		fmt.Println(result1)
		fmt.Println(result2)
		fmt.Printf("elapsed: %f ms\n", duration.Seconds()*1000.0)
	} else {
		result1, result2 := solve(data)

		fmt.Println(result1)
		fmt.Println(result2)
	}
}
